### PR TITLE
Final-Unit-Test1

### PR DESCRIPTION
**Pull request by olena11o**
Please review Test 1 for final homework.
Regards,
Olena

### CODE REVIEW

File: .eslintrc.js [hunk 1]
```diff
@@ -41,5 +42,17 @@ module.exports = {
...
+            functions: true,
+            classes: true,
+            variables: true,
+            allowNamedExports: false,
+        }],
+        'linebreak-style': [
+        'error', 'windows',
+        ],
+        'import/prefer-default-export': [
+            ["off" | "warn" | "error"],
```
Comment (EugenePetrik) [inline .eslintrc.js]: You should choose one of these options: or `off`, or `warn`, or `error`

File: .gitignore [hunk 2]
```diff
@@ -2,3 +2,15 @@ node_modules/
 /test-results/
 /playwright-report/
 /playwright/.cache/
+/test-results/
```
Comment (EugenePetrik) [inline .gitignore]: /test-results/
/playwright-report/
/playwright/.cache/

These options are already included in the gitignore file. You do not have to duplicate them.

File: src/pages/Inventory.page.js [hunk 3]
```diff
@@ -9,7 +9,21 @@ export class InventoryPage extends BaseSwagLabPage {
...
     get addItemToCartBtns() { return this.page.locator('[id^="add-to-cart"]'); }
 
+    get sortDropdown() { return this.page.locator('.product_sort_container'); }
+
     async addItemToCartById(id) {
         await this.addItemToCartBtns.nth(id).click();
     }
+
+    async inventoryItemsNames() {
+        return await this.inventoryItems.locator('[data-test="inventory-item-name"]').allTextContents();
```
Comment (EugenePetrik) [inline src/pages/Inventory.page.js]: `await` is redundant here, can be just

```
return this.inventoryItems.locator('[data-test="inventory-item-name"]').allTextContents();
```

File: src/tests/myFinalTests.spec.js [hunk 4]
```diff
@@ -0,0 +1,34 @@
+// @ts-check
+const { expect } = require('@playwright/test');
+const { test } = require('../fixture');
+const { currentOrderOnPage, sortArray, compareTwoArrays } = require('./utils');
+
+['az', 'za', 'lohi', 'hilo'].forEach((sortingType) => {
+    test.describe('Basic tests', () => {
+        test.beforeEach(async ({ loginPage, inventoryPage }) => {
+            await loginPage.navigate();
+            await loginPage.performLogin('standard_user', 'secret_sauce');
```
Comment (EugenePetrik) [inline src/tests/myFinalTests.spec.js]: it would be better to move user credentials to the separate module

File: src/tests/utils.js [hunk 5]
```diff
@@ -0,0 +1,42 @@
+async function currentOrderOnPage(sortingType, inventoryPage) {
```
Comment (EugenePetrik) [inline src/tests/utils.js]: Exports can be simplified 

```
export async function currentOrderOnPage(sortingType, inventoryPage) {}
```

File: e2e/example.spec.js [hunk 6]
```diff
@@ -0,0 +1,19 @@
+// @ts-check
```
Comment (EugenePetrik) [inline e2e/example.spec.js]: This file can be removed

File: tests/example.spec.js [hunk 7]
```diff
@@ -0,0 +1,19 @@
+// @ts-check
```
Comment (EugenePetrik) [inline tests/example.spec.js]: This file can be removed as well

File: src/tests/myFinalTests.spec.js [hunk 8]
```diff
@@ -0,0 +1,34 @@
+// @ts-check
```
Comment (EugenePetrik) [inline src/tests/myFinalTests.spec.js]: Let's rename the file name, i.e. `sortingOnTheInventoryPage.spec.js`.
This title would be more appropriate.

File: tests-examples/demo-todo-app.spec.js [hunk 9]
```diff
@@ -0,0 +1,449 @@
+// @ts-check
```
Comment (EugenePetrik) [inline tests-examples/demo-todo-app.spec.js]: This file can be removed too

File: src/tests/myFinalTests.spec.js [hunk 10]
```diff
@@ -0,0 +1,34 @@
...
+const { expect } = require('@playwright/test');
+const { test } = require('../fixture');
+const { currentOrderOnPage, sortArray, compareTwoArrays } = require('./utils');
+
+['az', 'za', 'lohi', 'hilo'].forEach((sortingType) => {
```
Comment (EugenePetrik) [inline src/tests/myFinalTests.spec.js]: Nice 👍

Comment (EugenePetrik) [reply src/tests/myFinalTests.spec.js]: I would improve this a little

```
[
{
sortingTitle: 'Z to A'.
sortingType: 'az'
},
{
title:
sortingType:
},
{
title: 'Low to High',
sortingType: 'lohi'
},
{
title:
sortingType:
}
].forEach(({sortingTitle, sortingType}) => {
test(`Test 1 - Verify sort by ${sortingTitle}`, async ({ inventoryPage }) => {});
}
```

In this way you will have more readable test names in the report

**Review by EugenePetrik — CHANGES_REQUESTED** [review]
Good job! Left a few comments

File: .eslintrc.js (hunk 1 above)
Comment (olena11o) [reply .eslintrc.js]: Updated

File: .gitignore (hunk 2 above)
Comment (olena11o) [reply .gitignore]: Updated

File: src/pages/Inventory.page.js (hunk 3 above)
Comment (olena11o) [reply src/pages/Inventory.page.js]: Updated

File: src/tests/utils.js (hunk 5 above)
Comment (olena11o) [reply src/tests/utils.js]: Updated

File: e2e/example.spec.js (hunk 6 above)
Comment (olena11o) [reply e2e/example.spec.js]: Updated

File: tests/example.spec.js (hunk 7 above)
Comment (olena11o) [reply tests/example.spec.js]: Updated

File: tests-examples/demo-todo-app.spec.js (hunk 9 above)
Comment (olena11o) [reply tests-examples/demo-todo-app.spec.js]: Updated

File: src/tests/myFinalTests.spec.js (hunk 8 above)
Comment (olena11o) [reply src/tests/myFinalTests.spec.js]: Updated

File: src/tests/myFinalTests.spec.js (hunk 10 above)
Comment (olena11o) [reply src/tests/myFinalTests.spec.js]: Updated

File: src/tests/myFinalTests.spec.js (hunk 4 above)
Comment (olena11o) [reply src/tests/myFinalTests.spec.js]: Updated

File: src/pages/Inventory.page.js (hunk 3 above)
Comment (hlazkov) [reply src/pages/Inventory.page.js]: So now no sence in `async` :) Its JFYI

File: src/tests/auth.setup.js [hunk 11]
```diff
@@ -0,0 +1,21 @@
+import { test as setup } from '@playwright/test';
+
+const authFile = 'playwright/.auth/user.json';
+
+setup('authenticate', async ({ page }) => {
+  // Perform authentication steps. Replace these actions with your own.
+  await page.goto('https://www.saucedemo.com/');
```
Comment (hlazkov) [inline src/tests/auth.setup.js]: Lets move `https://www.saucedemo.com/` to the baseUrl of config file

File: src/tests/utils.js [hunk 12]
```diff
@@ -0,0 +1,39 @@
...
+            sortedArray = arrayToSort.toSorted((a, b) => a - b);
+            break;
+        case 'hilo':
+            sortedArray = arrayToSort.toSorted((a, b) => b - a);
+            break;
+    };
+    return sortedArray;
+};
+
+export function compareTwoArrays(arrayActual, arrayExpected) {
```
Comment (hlazkov) [inline src/tests/utils.js]: Do we really need this function? default assertions should do this better

**Review by hlazkov — CHANGES_REQUESTED** [review]
Please recheck on comments provided

File: src/pages/Inventory.page.js (hunk 3 above)
Comment (olena11o) [reply src/pages/Inventory.page.js]: Updated

File: src/tests/auth.setup.js (hunk 11 above)
Comment (olena11o) [reply src/tests/auth.setup.js]: Updated

File: src/tests/utils.js (hunk 12 above)
Comment (olena11o) [reply src/tests/utils.js]: Updated

File: src/tests/auth.setup.js [hunk 13]
```diff
@@ -0,0 +1,22 @@
+import { test as setup } from '@playwright/test';
+
+const authFile = 'playwright/.auth/user.json';
+
+setup('authenticate', async ({ page }) => {
+  // Perform authentication steps. Replace these actions with your own.
+  await page.goto('');
```
Comment (EugenePetrik) [inline src/tests/auth.setup.js]: We can use page object methods here too


```
setup('authenticate', async ({ loginPage }) => {});
  await loginPage.navigate();
  ...
```

**Review by EugenePetrik — CHANGES_REQUESTED** [review]
Good job! Left a comment

File: src/tests/utils.js [hunk 14]
```diff
@@ -0,0 +1,29 @@
+export async function currentOrderOnPage(sortingType, inventoryPage) {
```
Comment (hlazkov) [inline src/tests/utils.js:1]: Lets move this file out of tests folder. I would rather have it in separate folder like `utils`

File: src/tests/auth.setup.js [hunk 15]
```diff
@@ -0,0 +1,38 @@
...
+//   // Wait until the page receives the cookies.
+//   //
+//   // Sometimes login flow sets cookies in the process of several redirects.
+//   // Wait for the final URL to ensure that the cookies are actually set.
+//  await page.waitForURL('inventory.html');
+
+
+//   // End of authentication steps.
+
+//   await page.context().storageState({ path: authFile });
```
Comment (hlazkov) [inline src/tests/auth.setup.js:37]: Lets clean up redundant lines here.

File: src/tests/sortingOnTheInventoryPage.spec.js [hunk 16]
```diff
@@ -0,0 +1,46 @@
...
+    },
+    {
+    sortingTitle: 'Low to High',
+    sortingType: 'lohi',
+    },
+    {
+    sortingTitle: 'High to Low',
+    sortingType: 'hilo',
+    }
+    ]
```
Comment (hlazkov) [inline src/tests/sortingOnTheInventoryPage.spec.js:22]: To make this more readable, you could update this array with the proper spacing.

File: src/tests/sortingOnTheInventoryPage.spec.js [hunk 17]
```diff
@@ -0,0 +1,46 @@
...
+            const sortExpected = sortArray(sortingType, orderOnPage);
+            
+            // Get actual result on page after sorting
+            await inventoryPage.sortBy(sortingType);
+            const sortActual = await currentOrderOnPage(sortingType, inventoryPage);
+
+            // Compare actual and expected items lists    
+            expect(sortActual.length).toEqual(sortExpected.length);
+    
+            //const compareResult = compareTwoArrays(sortActual, sortExpected);
```
Comment (hlazkov) [inline src/tests/sortingOnTheInventoryPage.spec.js:39]: Lets remove all the code we dont need.

**Review by hlazkov — APPROVED** [review]
Good job! I've left some comments below to improve this code, I hope you will follow these advices before merge :)

File: src/tests/auth.setup.js (hunk 13 above)
Comment (olena11o) [reply src/tests/auth.setup.js]: Updated

File: src/tests/utils.js (hunk 14 above)
Comment (olena11o) [reply src/tests/utils.js:1]: Updated

File: src/tests/auth.setup.js (hunk 15 above)
Comment (olena11o) [reply src/tests/auth.setup.js:37]: Updated

File: src/tests/sortingOnTheInventoryPage.spec.js (hunk 17 above)
Comment (olena11o) [reply src/tests/sortingOnTheInventoryPage.spec.js:39]: Updated

**Pull request merged**